Add null checks to `Hash.serialize` and `Hash.deserialize`

### Problem

This is a follow up for the investigation of a failure happened during 0.48 -> 0.48.1 upgrade. 

The following exception happened on one of the node. 
```
2024-04-17 00:08:31.898 52       ERROR EXCEPTION        <main> MerkleUtils: Async tree hashing failed
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot read the array length because "input" is null
	at com.swirlds.common.merkle.hash.FutureMerkleHash.rethrowException(FutureMerkleHash.java:83) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:93) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:30) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.utility.MerkleUtils.rehashTree(MerkleUtils.java:67) [swirlds-common-0.48.1.jar:?]
	at com.swirlds.platform.state.signed.StartupStateUtils.loadStateFile(StartupStateUtils.java:520) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.state.signed.StartupStateUtils.loadLatestState(StartupStateUtils.java:473) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.state.signed.StartupStateUtils.loadStateFile(StartupStateUtils.java:223) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.state.signed.StartupStateUtils.getInitialState(StartupStateUtils.java:150) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.PlatformBuilder.build(PlatformBuilder.java:235) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.Browser.launchUnhandled(Browser.java:189) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.Browser.launch(Browser.java:128) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.Browser.parseCommandLineArgsAndLaunch(Browser.java:103) [swirlds-platform-core-0.48.1.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:92) [swirlds-platform-core-0.48.1.jar:?]
Caused by: java.lang.NullPointerException: Cannot read the array length because "input" is null
	at java.base/java.security.MessageDigest.update(MessageDigest.java:357) ~[?:?]
	at com.swirlds.common.crypto.HashBuilder.update(HashBuilder.java:172) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.crypto.internal.MerkleInternalDigestProvider.handleItem(MerkleInternalDigestProvider.java:79) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.crypto.internal.MerkleInternalDigestProvider.compute(MerkleInternalDigestProvider.java:91) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.crypto.internal.MerkleCryptoEngine.digestSync(MerkleCryptoEngine.java:92) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.crypto.internal.MerkleCryptoEngine.digestSync(MerkleCryptoEngine.java:116) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.crypto.MerkleCryptography.digestSync(MerkleCryptography.java:55) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.hashSubtree(MerkleHashBuilder.java:213) ~[swirlds-common-0.48.1.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.lambda$createHashingRunnable$1(MerkleHashBuilder.java:173) ~[swirlds-common-0.48.1.jar:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
```

Even though the state files were available, the issue couldn't be reproduced. The exception above could be more specific and provide additional information like parent class, child class and child route. Unfortunately, we didn't take into account the possibility of `Hash#value` being null. Strictly speaking, it never should be null. Currently, there is a check for that in the constructor but not in the serializer/deserializer methods. These methods need null checks. 



### Solution

- Update `handleItem` method to check for `hash.value == null` condition. 
- Add assert to `serialize` method
- Add check for null to `deserialize` method

### Alternatives

_No response_

--- a/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/crypto/Hash.java
+++ b/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/crypto/Hash.java
@@ -192,6 +192,8 @@ public class Hash implements Comparable<Hash>, SerializableWithKnownLength, Seri
      */
     @Override
     public void serialize(final SerializableDataOutputStream out) throws IOException {
+        assert digestType != null;
+        assert value != null;
         out.writeInt(digestType.id());
         out.writeByteArray(value);
     }
@@ -214,6 +216,10 @@ public class Hash implements Comparable<Hash>, SerializableWithKnownLength, Seri
 
         this.digestType = digestType;
         this.value = in.readByteArray(digestType.digestLength());
+
+        if (value == null) {
+            throw new BadIOException("Invalid hash value read from the stream");
+        }
     }
 
     /**

--- a/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/crypto/Hash.java
+++ b/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/crypto/Hash.java
@@ -18,6 +18,7 @@ package com.swirlds.common.crypto;
 
 import static com.swirlds.common.utility.CommonUtils.hex;
 import static com.swirlds.common.utility.Mnemonics.generateMnemonic;
+import static java.util.Objects.requireNonNull;
 
 import com.hedera.pbj.runtime.io.buffer.Bytes;
 import com.swirlds.common.io.SerializableWithKnownLength;
@@ -192,8 +193,8 @@ public class Hash implements Comparable<Hash>, SerializableWithKnownLength, Seri
      */
     @Override
     public void serialize(final SerializableDataOutputStream out) throws IOException {
-        assert digestType != null;
-        assert value != null;
+        requireNonNull(digestType, "digestType");
+        requireNonNull(value, "value");
         out.writeInt(digestType.id());
         out.writeByteArray(value);
     }

--- a/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/crypto/internal/MerkleInternalDigestProvider.java
+++ b/platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/crypto/internal/MerkleInternalDigestProvider.java
@@ -65,7 +65,7 @@ public class MerkleInternalDigestProvider
         for (int index = 0; index < childHashes.size(); index++) {
             final Hash childHash = childHashes.get(index);
 
-            if (childHash == null) {
+            if (childHash == null || childHash.getValue() == null) {
                 final MerkleNode childNode = node.getChild(index);
                 final String msg = String.format(
                         "Child has an unexpected null hash "

